File: 3Sum/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        
        for i,n in enumerate(nums):
            if n>0: break
            if i>=len(nums)-2: break
            if i>0 and n==nums[i-1]: continue
            l,r = i+1, len(nums)-1
            while l<r:
                s = nums[l]+nums[r]
                if s+n==0:
                    ans.append([n, nums[l], nums[r]])
                    l+=1
                    r-=1
                elif s+n<0:
                    l+=1
                elif s+n>0:
                    r-=1
                while i+1<l<len(nums)-1 and nums[l]==nums[l-1]: l+=1
                while i+1<r<len(nums)-1 and nums[r]==nums[r+1]: r-=1
        return ans

File: 3Sum/3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        // System.out.println(Arrays.toString(nums));

        int index=0;
        int m,n;
        List list = new ArrayList<>();
        int target=0;
        if(nums.length<3) return list;
        
        while(nums[index]<=0){
            if(index>nums.length-3) break;
            m=index+1;
            n=nums.length-1;
            target = -1*nums[index];

            while(n>m){
                if(nums[n]+nums[m]>target){
                    do{
                        n--;
                    }while(nums[n]==nums[n+1] && n>m);
                }
                else if(nums[n]+nums[m]<target){
                    do{
                        m++;
                    }while(nums[m]==nums[m-1]&& m<n);
                }
                else{
                    List templist = new ArrayList<>();
                    templist.add(nums[index]);
                    templist.add(nums[m]);
                    templist.add(nums[n]);
                    list.add(templist);
                    do{
                        m++;
                    }while(nums[m]==nums[m-1] && m<n);
                }
            }

            do{
                index++;
            }while(nums[index]==nums[index-1] && index<m);
            if(index>nums.length-3) break;
        }

        return list;
    }
}